File: javascript/packages/core/src/models/invoice.js
/** Invoice issued for a completed order. */

import { Money } from "../money.js";

export class Invoice {
    /**
     * @param {number} id
     * @param {number} repairOrderId
     * @param {Money} total
     */
    constructor(id, repairOrderId, total) {
        this.id = id;
        this.repairOrderId = repairOrderId;
        this.total = total;
        this.paid = false;
    }

    /** @returns {boolean} */
    markPaid() {
        if (this.paid) {
            return false;
        }

        this.paid = true;

        return true;
    }

    /** @returns {Money} */
    outstanding() {
        return this.paid ? Money.ZERO : this.total;
    }
}

File: rust/crates/atelier-core/src/models/invoice.rs
//! Invoice issued for a completed order.

use crate::money::Money;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Invoice {
    pub id: u32,
    pub repair_order_id: u32,
    pub total: Money,
    pub paid: bool,
}

impl Invoice {
    pub fn new(id: u32, repair_order_id: u32, total: Money) -> Self {
        Self {
            id,
            repair_order_id,
            total,
            paid: false,
        }
    }

    pub fn mark_paid(&mut self) -> bool {
        if self.paid {
            return false;
        }

        self.paid = true;

        true
    }

    pub fn outstanding(&self) -> Money {
        if self.paid { Money::ZERO } else { self.total }
    }
}
